sdk/core/azure-core-amqp/src/models/amqp_properties.cpp

```diff
@@ -0,0 +1,442 @@
+        throw std::runtime_error("Could not set to"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.Subject.HasValue())
+    {
+      if (properties_set_subject(returnValue.get(), properties.Subject.Value().data()))
+      {
+        throw std::runtime_error("Could not set subject");
+      }
+    }
+
+    if (properties.ReplyTo.HasValue())
+    {
+      if (properties_set_reply_to(returnValue.get(), properties.ReplyTo.Value()))
+      {
+        throw std::runtime_error("Could not set reply to"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.ContentType.HasValue())
+    {
+      if (properties_set_content_type(returnValue.get(), properties.ContentType.Value().data()))
+      {
+        throw std::runtime_error("Could not set content type"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.ContentEncoding.HasValue())
+    {
+      if (properties_set_content_encoding(
+              returnValue.get(), properties.ContentEncoding.Value().data()))
+      {
+        throw std::runtime_error("Could not set content type");
+      }
+    }
+
+    if (properties.AbsoluteExpiryTime.HasValue())
+    {
+      auto timeStamp{std::chrono::duration_cast<std::chrono::milliseconds>(
+          properties.AbsoluteExpiryTime.Value().time_since_epoch())};
+
+      if (properties_set_absolute_expiry_time(returnValue.get(), timeStamp.count()))
+      {
+        throw std::runtime_error("Could not set absolute expiry time"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.CreationTime.HasValue())
+    {
+      auto timeStamp{std::chrono::duration_cast<std::chrono::milliseconds>(
+          properties.CreationTime.Value().time_since_epoch())};
+
+      if (properties_set_creation_time(returnValue.get(), timeStamp.count()))
+      {
+        throw std::runtime_error("Could not set absolute expiry time"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.GroupId.HasValue())
+    {
+      if (properties_set_group_id(returnValue.get(), properties.GroupId.Value().data()))
+      {
+        throw std::runtime_error("Could not set group id"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.GroupSequence.HasValue())
+    {
+      if (properties_set_group_sequence(returnValue.get(), properties.GroupSequence.Value()))
+      {
+        throw std::runtime_error("Could not set group sequence"); // LCOV_EXCL_LINE
+      }
+    }
+
+    if (properties.ReplyToGroupId.HasValue())
+    {
+      if (properties_set_reply_to_group_id(
+              returnValue.get(), properties.ReplyToGroupId.Value().data()))
+      {
+        throw std::runtime_error("Could not set reply-to group id");
+      }
+    }
+
+    return returnValue;
+  }
+
+  namespace {
+
+    template <typename T> bool CompareNullable(T const& left, T const& right)
+    {
+      if (left.HasValue() != right.HasValue())
+      {
+        return false;
+      }
+      if (left.HasValue())
+      {
+        return left.Value() == right.Value();
+      }
+      return true;
```

Comment:
That's a matter of choice, but I'd wrote it as `return left.HasValue() == right.HasValue() && (!left.HasValue() || left.Value() == right.Value())`.

Reply:
I suspect the compiler will generate exactly the same code for both of these constructs.

And personally, I find the construction I am using to be easier to understand than yours (even though they are functionally identical) - I'm not a huge fan of deeply nested boolean conditionals.

Reply:
> And personally, I find the construction I am using to be easier to understand than yours (even though they are functionally identical) - I'm not a huge fan of deeply nested Boolean conditionals.

+1